diff --git a/plugin/auth_ldap/CMakeLists.txt b/plugin/auth_ldap/CMakeLists.txt
index d9649b30..983e0db0 100644
--- a/plugin/auth_ldap/CMakeLists.txt
+++ b/plugin/auth_ldap/CMakeLists.txt
@@ -13,6 +13,7 @@
 # along with this program; if not, write to the Free Software
 # Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA

+IF(WITH_AUTHENTICATION_LDAP)
 IF(WITH_LDAP)
   INCLUDE (CheckLibraryExists)
   CHECK_LIBRARY_EXISTS(ldap ldap_initialize "" HAVE_LDAP)
@@ -45,3 +46,4 @@ IF(WITH_LDAP)
   ENDIF(UNIX)

 ENDIF(WITH_LDAP)
+ENDIF()

