diff --git a/percona-xtradb-cluster-galera/CMakeLists.txt b/percona-xtradb-cluster-galera/CMakeLists.txt
index 98c85a7b..635eadc5 100644
--- a/percona-xtradb-cluster-galera/CMakeLists.txt
+++ b/percona-xtradb-cluster-galera/CMakeLists.txt
@@ -97,6 +97,11 @@ include(cmake/maintainer_mode.cmake)
 
 include(CTest)
 enable_testing()
+
+if(DEFINED WITH_PERFSCHEMA_STORAGE_ENGINE)
+  add_definitions(-DHAVE_PSI_INTERFACE=1)
+endif()
+
 add_subdirectory(galerautils)
 add_subdirectory(gcomm)
 add_subdirectory(gcache)
diff --git a/percona-xtradb-cluster-galera/galera/src/CMakeLists.txt b/percona-xtradb-cluster-galera/galera/src/CMakeLists.txt
index 9d14e931..62e8a2de 100644
--- a/percona-xtradb-cluster-galera/galera/src/CMakeLists.txt
+++ b/percona-xtradb-cluster-galera/galera/src/CMakeLists.txt
@@ -19,6 +19,7 @@ add_library(galera STATIC
   galera_info.cpp
   replicator.cpp
   ist.cpp
+  socket_watchdog.cpp
   gcs_dummy.cpp
   saved_state.cpp
   replicator_smm.cpp
@@ -26,6 +27,8 @@ add_library(galera STATIC
   replicator_smm_stats.cpp
   )
 
+set_source_files_properties(socket_watchdog.cpp PROPERTIES COMPILE_FLAGS -std=c++11)
+
 target_include_directories(galera
   PRIVATE
   ${CMAKE_SOURCE_DIR}/wsrep/src
diff --git a/percona-xtradb-cluster-galera/gcs/src/CMakeLists.txt b/percona-xtradb-cluster-galera/gcs/src/CMakeLists.txt
index ab9f7bb8..b6bad75f 100644
--- a/percona-xtradb-cluster-galera/gcs/src/CMakeLists.txt
+++ b/percona-xtradb-cluster-galera/gcs/src/CMakeLists.txt
@@ -55,8 +55,12 @@ target_link_libraries(gcs gcomm gcache)
 add_library(gcs4garb STATIC ${GCS_SOURCES})
 
 target_compile_definitions(gcs4garb
-  PRIVATE
+  PUBLIC
   -DGCS_FOR_GARB
+  )
+
+target_compile_definitions(gcs4garb
+  PRIVATE
   -DGALERA_LOG_H_ENABLE_CXX
   -DGCS_USE_GCOMM
   )
