diff --git a/percona-xtradb-cluster-galera/garb/CMakeLists.txt b/percona-xtradb-cluster-galera/garb/CMakeLists.txt
index 40b784ce..188185f1 100644
--- a/percona-xtradb-cluster-galera/garb/CMakeLists.txt
+++ b/percona-xtradb-cluster-galera/garb/CMakeLists.txt
@@ -20,6 +20,7 @@ target_compile_definitions(garbd
   PRIVATE
   -DGALERA_VER="${GALERA_VERSION}"
   -DGALERA_REV="${GALERA_REVISION}"
+  -DGCS_FOR_GARB
   )

 # TODO: Fix.