diff --git a/percona-xtradb-cluster-galera/CMakeLists.txt b/percona-xtradb-cluster-galera/CMakeLists.txt
index f0a8013e..eef14cf9 100644
--- a/percona-xtradb-cluster-galera/CMakeLists.txt
+++ b/percona-xtradb-cluster-galera/CMakeLists.txt
@@ -99,6 +99,11 @@ include(cmake/maintainer_mode.cmake)
 
 include(CTest)
 enable_testing()
+
+if(DEFINED WITH_PERFSCHEMA_STORAGE_ENGINE)
+  add_definitions(-DHAVE_PSI_INTERFACE=1)
+endif()
+
 add_subdirectory(galerautils)
 add_subdirectory(gcomm)
 add_subdirectory(gcache)
diff --git a/percona-xtradb-cluster-galera/gcs/src/CMakeLists.txt b/percona-xtradb-cluster-galera/gcs/src/CMakeLists.txt
index 5bbe028c..8b6b9fff 100644
--- a/percona-xtradb-cluster-galera/gcs/src/CMakeLists.txt
+++ b/percona-xtradb-cluster-galera/gcs/src/CMakeLists.txt
@@ -58,8 +58,12 @@ target_link_libraries(gcs gcomm gcache)
 add_library(gcs4garb STATIC ${GCS_SOURCES})
 
 target_compile_definitions(gcs4garb
-  PRIVATE
+  PUBLIC
   -DGCS_FOR_GARB
+  )
+
+target_compile_definitions(gcs4garb
+  PRIVATE
   -DGALERA_LOG_H_ENABLE_CXX
   -DGCS_USE_GCOMM
   )
